videostore/indexer.go

```diff
@@ -0,0 +1,406 @@
+	TotalDurationMs int64
+	SegmentCount    int
+	Ranges          []timeRange // List of contiguous time ranges
+}
+
+// GetStorageState calculates the current storage state from the index.
+// This function acquires the indexer's mutex.
+func (ix *Indexer) GetStorageState() (StorageState, error) {
+	ix.mu.Lock()
+	defer ix.mu.Unlock()
+
+	var state StorageState
+
+	// Query all segments, ordered by start time
+	query := fmt.Sprintf(`
+	SELECT start_time_unix, duration_ms, size_bytes
+	FROM %s
+	ORDER BY start_time_unix ASC
+	`, segmentsTableName)
+
+	rows, err := ix.db.Query(query)
+	if err != nil {
+		return StorageState{}, fmt.Errorf("failed to query segments for state: %w", err)
+	}
+	defer rows.Close()
+
+	var ranges []timeRange
+	var currentRange *timeRange
+	const slopDuration = 5 * time.Second
+
+	for rows.Next() {
+		var startTimeUnix int64
+		var durationMs sql.NullInt64
+		var sizeBytes sql.NullInt64
+
+		if err := rows.Scan(&startTimeUnix, &durationMs, &sizeBytes); err != nil {
+			return StorageState{}, fmt.Errorf("failed to scan segment row: %w", err)
+		}
+
+		state.SegmentCount++
+		if durationMs.Valid {
+			state.TotalDurationMs += durationMs.Int64
+		} else {
+			ix.logger.Warnw("segment found with NULL duration in index", "start_time_unix", startTimeUnix)
+			continue
+		}
+		if sizeBytes.Valid {
+			state.TotalSizeBytes += sizeBytes.Int64
+		}
+
+		segmentStart := time.Unix(startTimeUnix, 0)
+		segmentEnd := segmentStart.Add(time.Duration(durationMs.Int64) * time.Millisecond)
+
+		if currentRange == nil {
+			currentRange = &timeRange{Start: segmentStart, End: segmentEnd}
+		} else {
+			if segmentStart.After(currentRange.End.Add(slopDuration)) {
+				ranges = append(ranges, *currentRange)
+				currentRange = &timeRange{Start: segmentStart, End: segmentEnd}
+			} else {
+				currentRange.End = segmentEnd
+			}
+		}
+	}
+	if currentRange != nil {
+		ranges = append(ranges, *currentRange)
+	}
+
+	if err := rows.Err(); err != nil {
+		return StorageState{}, fmt.Errorf("error iterating over all segments query result: %w", err)
+	}
+
+	state.Ranges = ranges
+	return state, nil
+}
+
+// GetSegmentsAscTime retrieves all indexed segments in ascending order by start time.
+// Returns a slice of SegmentMetadata containing file path and size.
+// This function acquires the indexer's mutex.
+func (ix *Indexer) GetSegmentsAscTime() ([]SegmentMetadata, error) {
+	ix.mu.Lock()
+	defer ix.mu.Unlock()
+
+	query := fmt.Sprintf(`
+	SELECT file_path, size_bytes
+	FROM %s
+	ORDER BY start_time_unix ASC
+	`, segmentsTableName)
+
+	rows, err := ix.db.Query(query)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query all segments: %w", err)
+	}
+	defer rows.Close()
+
+	var segments []SegmentMetadata
+
+	for rows.Next() {
+		var filePath string
+		var sizeBytes sql.NullInt64
+
+		if err := rows.Scan(&filePath, &sizeBytes); err != nil {
+			ix.logger.Errorw("failed to scan segment row during full query", "error", err)
+			continue
```

Comment:
why not return this as an error?

Reply:
It is so one single problematic row doesn't affect us returning *something* to the user.

Reply:
but then we are lying and returning incomplete data.

If we don't know specifically why this error would happen for one row and not all rows, lets log which row we failed on and fail the entire request if we hit an error.

Reply:
Changed to error return